src/Interpreters/Context.cpp

```diff
@@ -686,15 +686,15 @@ void Context::setUser(const String & name, const String & password, const Poco::
 
 #if defined(ARCADIA_BUILD)
     /// This is harmful field that is used only in foreign "Arcadia" build.
-    client_info.current_password = password;
+    client_info.current_password = password.value_or("");
 #endif
 
     auto new_user_id = getAccessControlManager().find<User>(name);
     std::shared_ptr<const ContextAccess> new_access;
     if (new_user_id)
     {
         new_access = getAccessControlManager().getContextAccess(*new_user_id, {}, true, settings, current_database, client_info);
-        if (!new_access->isClientHostAllowed() || !new_access->isCorrectPassword(password))
+        if (!new_access->isClientHostAllowed() || (password && !new_access->isCorrectPassword(*password)))
```

Comment:
Should we check that the client host is allowed if that's a secondary query? It seems we shouldn't.